python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+        pass
+    worker.register_activity(nested_activity)
+    # Fails because it's not callable
+    worker.register_activity(MyActivities())
+
+    # Registers only "activity1"
+    worker.register_activities(MyActivities())
+    # As if calling register_activity on each
+    worker.register_activities([another_activity, MyCallableClass()])
+    # As if calling register_activity with a custom name on each
+    worker.register_activities({'activity_name': simple_activity})
+
+    # Fails, has no @activity.run method
+    worker.register_activities(MyCallableClass())
+```
+
+Notes:
+
+* We use `@activity.run` from `temporalio.activity` instead of just `@activity` from because:
+  * We want `activity` to be a module because we want it usually imported as `from temporalio import activity` so
+    top-level calls like `activity.info()` read well
+  * When we get to workflows, we will do similar with `@workflow.run` vs `@workflow.signal` vs `@workflow.query`, etc
+    and `workflow.info()`
+* Cancellation is task cancellation
+  * Do we need to differentiate worker shutdown from server-side cancellation?
+  * So we need an explicit future they can wait on for activity cancellation?
+
+### Data Converters
+
+Exported from `temporalio` (but probably in `temporalio/converter/converter.py`, exported from `temporalio.converter`
+then exported from `temporalio`):
+
+```python
+class DataConverter(ABC):
+    # This can be sync or async. Always takes a list. Can return a single
+    # payload or a list of payloads with the same size as the input list or None
+    # if the value cannot be converted with this converter. Any other result
+    # causes an error. Any error raised is bubbled.
+    @abstractmethod
+    async def encode(self, *values):
+        pass
+
+    # This can be sync or async. Always takes a list. Can return a single value
+    # or a list of values with the same size as the input list or None if the
+    # value cannot be converted with this converter. Any other result causes an
+    # error. Any error raised is bubbled.
+    @abstractmethod
+    async def decode(self, *payloads):
+        pass
+
+
+class CompositeDataConverter(DataConverter):
+    def __init__(self, converters) -> None:
+        pass
+
+    async def encode(self, *values):
+        pass
+
+    async def decode(self, *payloads):
+        pass
+
+
+DataConverter.default = CompositeDataConverter(
+    [
+        BinaryNullDataConverter(),
+        BinaryPlainDataConverter(),
+        JSONProtoDataConverter(),
+        BinaryProtoDataConverter(),
+        JSONPlainDataConverter(),
+    ]
+)
+```
+
+Notes:
+
+* May be a bit difficult to allow async and non-async converter w/ same base but we can
+  * Ref https://stackoverflow.com/questions/47555934/how-require-that-an-abstract-method-is-a-coroutine
+* This is a combination of the otherwise separate concepts of "DataConverter" and "PayloadConverter" from other SDKs
+  since the extra abstraction seems unnecessary, is that ok?
```

Comment:
I'm not sure what the point of the distinction is either.

I do think we eventually want separate bytes-to-bytes processors from the serializers, but different story.

Reply:
We don't have to do that for Python, the same interface can be used for both.

Reply:
I think we can unify the 2 concepts into a single class, I think `PayloadConverter` is a better name because payloads are a known Temporal term.

Reply:
I think "Data Converter" is a known Temporal term too, but I also prefer `PayloadConverter`. I've been looking around, and technically a `PayloadEncoder` + `PayloadDecoder` are more apt terms, but we combine them and there's no settled name in the ecosystem for such a combination. So I will switch to `PayloadConverter`.

Reply:
Changed to payload converter